pilot/pkg/model/network.go

```diff
@@ -510,22 +532,22 @@ func (n *networkGatewayNameCache) refreshAndNotify(name string) func() {
 }
 
 // resolve gets all the A and AAAA records for the given name
-func (n *networkGatewayNameCache) resolve(name string) ([]string, time.Duration) {
+func (n *networkGatewayNameCache) resolve(name string) (out []string, d time.Duration, err error) {
 	ttl := uint32(math.MaxUint32)
-	var out []string
+	errs := istiomultierror.New()
 
 	var mu sync.Mutex
 	var wg sync.WaitGroup
 	doResolve := func(dnsType uint16) {
 		defer wg.Done()
 
 		res := n.client.Query(new(dns.Msg).SetQuestion(dns.Fqdn(name), dnsType))
-		if len(res.Answer) == 0 {
-			return
-		}
 
 		mu.Lock()
 		defer mu.Unlock()
+		if res.Rcode == dns.RcodeServerFailure {
+			errs = multierror.Append(errs, fmt.Errorf("upstream dns failure, qtype: %v", dnsType))
```

Comment:
Appending in concurrent go-routines is not safe

Reply:
>Appending in concurrent go-routines is not safe

This is protected by the mutex